clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -2989,6 +3000,24 @@ class CIRCmpThreeWayOpLowering
   }
 };
 
+class BaseClassAddrLowering
+    : public mlir::OpConversionPattern<mlir::cir::BaseClassAddrOp> {
+
+  using mlir::OpConversionPattern<
+      mlir::cir::BaseClassAddrOp>::OpConversionPattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::BaseClassAddrOp op, OpAdaptor adaptor,
+                  mlir::ConversionPatternRewriter &rewriter) const override {
+    auto llvmSrcVal = adaptor.getOperands().front();
+    auto llvmDstTy =
+        getTypeConverter()->convertType(op.getBaseAddr().getType());
+    rewriter.replaceOpWithNewOp<mlir::LLVM::BitcastOp>(op, llvmDstTy,
+                                                       llvmSrcVal);
```

Comment:
`cir.base_class_addr` is not simply a bitcast. It is expected to perform some pointer arithmetic to convert an object pointer to a base subobject pointer.

Reply:
Thanks for the tip. So, like you mentioned, `BaseClassAddrOp` needs to be more abstract in its syntax. For instance, we should also include the baseoffset values for casting to base classes in this attribute. And instead of generating them directly in CIRGen, they should all be handled together during the lowering phase. That's what I've got in mind. Any other situations?
 I removed this in the PR, and perhaps it can be implemented separately later on.

Reply:
At least 2 approaches exist that provides necessary information to `cir.base_class_addr` and allows proper lowering of it:

1. Attach `cir.base_class_addr` with a new attribute, e.g. `#cir.base_class_addr_info`, that contains the base-to-derived offsets. During lowering prepare, the offsets are extracted from the attribute and used for lowering the operation.
2. Attach `cir.base_class_addr` with the AST node corresponds to the derived-to-base cast expression. During lowering prepare, execute any necessary AST queries to find the offset for lowering the operation.

An example of the first approach can be found at the implementation of the `cir.dyn_cast` operation. The operation carries a `#cir.dyn_cast_info` attribute that contains all necessary ABI-specific information for lowering the operation during lowering prepare. While the 2nd approach allows a clearer assembly syntax.

As a side note, `cir.base_class_addr` should also handle derived-to-virtual-base casts whose offsets cannot be known statically. Thus we need at least 2 offsets to do the conversion job: 1) the address point in the vtable that corresponds to the vbase offset entry; 2) a fixed offset from the derived object (or vbase subobject) to the target base object.

Reply:
@Lancern very good explanation!

@Laity000 I suggest you start simple: only handle static offsets for now and assert when that's not the case. Please let's build this incrementally. Both (1) and (2) seems fine to me, but I'm inclined for (1) given that we already raise a bunch of vtable information and we have a use for that in mind in the future.